cli/command_repository_upgrade.go

```diff
@@ -0,0 +1,291 @@
+	return func(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+		if !c.skip {
+			if err := act(ctx, rep); err != nil {
+				// Explicitly skip all stages on error because tests do not
+				// skip/exit on error. Tests override os.Exit() that prevents
+				// running rest of the phases until we set the skip flag here.
+				// This flag is designed for testability and also to support
+				// rollback.
+				c.skip = true
+				return err
+			}
+		}
+
+		return nil
+	}
+}
+
+// setLockIntent is an upgrade phase which sets the upgrade lock intent with
+// desired parameters.
+func (c *commandRepositoryUpgrade) setLockIntent(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+	if c.ioDrainTimeout < repo.DefaultRepositoryBlobCacheDuration && !c.force {
+		return errors.Errorf("minimum required io-drain-timeout is %s", repo.DefaultRepositoryBlobCacheDuration)
+	}
+
+	now := rep.Time()
+	mp := rep.ContentReader().ContentFormat().MutableParameters
+	openOpts := c.svc.optionsFromFlags(ctx)
+	l := &repo.UpgradeLockIntent{
+		OwnerID:                openOpts.UpgradeOwnerID,
+		CreationTime:           now,
+		AdvanceNoticeDuration:  c.advanceNoticeDuration,
+		IODrainTimeout:         c.ioDrainTimeout,
+		StatusPollInterval:     c.statusPollInterval,
+		Message:                fmt.Sprintf("Upgrading from format version %d -> %d", mp.Version, content.MaxFormatVersion),
+		MaxPermittedClockDrift: MaxPermittedClockDrift(),
+	}
+
+	// Update format-blob and clear the cache.
+	// This will fail if we have already upgraded.
+	l, err := rep.SetUpgradeLockIntent(ctx, *l)
+	if err != nil {
+		return errors.Wrap(err, "error setting the upgrade lock intent")
+	}
+	// we need to reopen the repository after this point
+
+	locked, _ := l.IsLocked(now)
+	if l.AdvanceNoticeDuration != 0 && !locked {
+		upgradeTime := l.UpgradeTime()
+		log(ctx).Infof("Repository upgrade advance notice has been set, you must come back and perform the upgrade at %s.",
+			upgradeTime)
+
+		c.skip = true
+
+		return nil
+	}
+
+	log(ctx).Infof("Repository upgrade lock intent has been placed.")
+
+	return nil
+}
+
+// drainOrCommit is the upgrade CLI phase that will actually wait for all the
+// clients to be drained out of the upgrade quorum. This phase will cause all other phases to be
+// skipped until the lock is fully established.
+func (c *commandRepositoryUpgrade) drainOrCommit(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+	cf := rep.ContentReader().ContentFormat()
+	if cf.MutableParameters.EpochParameters.Enabled {
+		log(ctx).Infof("Repository indices have already been migrated to the epoch format, no need to drain other clients")
+
+		l, err := rep.GetUpgradeLockIntent(ctx)
+		if err != nil {
+			return errors.Wrap(err, "failed to get upgrade lock intent")
+		}
+
+		if l.AdvanceNoticeDuration == 0 {
```

Comment:
I do not understand the behavior here. If no (advance) notice was set, then no client draining is performed? that seems incorrect, unless I'm missing something here.

Reply:
the if conditions wrapping this and the the advance notice block here is saying this:

if we have already upgraded (to version 2) and no advance notice was set then let the upgrade continue